Fix: AttributeError for some hypergeometric expr in hypersimp

Harikrishna-Srinivasan opened the pull request:
#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes: #27901 

#### Brief description of what is fixed or changed
Changed `hypersimp` to correctly simplify some more hypergeometric expressions like `binomial`.

#### Other comments
Added test cases for its working.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

haru-44 committed:
Fix `FURB132`

haru-44 committed:
Fix `FURB136`

Fix for ruff rule `FURB152`

No modifications have been made to the test code.

Fix `FURB154`

Fix `FURB156`

Harikrishna-Srinivasan committed:
Fix: AttributeError for some hypergeometric expr in hypersimp

Harikrishna-Srinivasan commented:
Does this PR resolve the issue #27901 completely?

oscarbenjamin commented:
> Does this PR resolve the issue #27901 completely?

No. The first thing that is needed is that `rsolve_hyper` should check whether `hypersimp` returns None.

As for changing `hypersimp` to call `gammasimp` this is something that needs to be thought about carefully by someone looking at what the two functions do why this case doesn't already work, why it is that gammasimp affects that and so on. Just opening a PR with that change and asking if it fixes the issue isn't really helpful.

Harikrishna-Srinivasan committed:
Bug: None Check added to rsolve_hyper

Harikrishna-Srinivasan commented:
> As for changing `hypersimp` to call `gammasimp` this is something that needs to be thought about carefully by someone looking at what the two functions do why this case doesn't already work, why it is that gammasimp affects that and so on. Just opening a PR with that change and asking if it fixes the issue isn't really helpful.

Previously, `hypersimp` used `g.rewrite(gamma)` and then simplified.

Since, `gamma` func. are expanded using `expand_func` (but `factorial` aren't), I changed it to `gammasimp` (combining both rewrite & simplification)

I didn't add `None` check in rsolve_hyper previously, as this case will not happen, if the terms are hypergeometric.
But for safer side, now added.

oscarbenjamin commented on sympy/simplify/simplify.py:
```diff
@@ -313,11 +313,13 @@ def hypersimp(f, k):
     g = f.subs(k, k + 1) / f
 
-    g = g.rewrite(gamma)
+    g = gammasimp(g)
```
I'm wondering whether it would be better to have `g = cancel(expand_func(g.rewrite(gamma)))` here.

Does `gammasimp` do anything that is better than that for any of the examples?

The `gammasimp` function is quite complicated so it is not completely clear to me if everything it does is really needed or helpful here.

Harikrishna-Srinivasan replied:
I see that `cancel(expand_func(g.rewrite(gamma))).factor()` performs better than `gammasimp` & also passes most test cases.
So, I'm changing it to `cancel(expand_func(g.rewrite(gamma))).factor()` instead.
Thanks for pointing this. (Initially I thought that `cancel()` & `factor()` might slow down)

Harikrishna-Srinivasan replied:
See https://github.com/sympy/sympy/pull/27931#issuecomment-2799585589

oscarbenjamin commented:
I'm not going to accept this change to `hypersimp` until I understand what is happening. You need to give more explanation or this pull request is not going to be merged.

What are examples when `factor` is needed?

What does "performs poorer" mean? Do you mean that it is slower or that it does not cover all cases? If you mean that it is slower then show timings. If it does not cover all cases then give examples.

At least under some conditions `gammasimp` calls `cancel` or `factor`. I don't understand in general what it does because it is quite complex though.

Harikrishna-Srinivasan commented:
I'm really sorry that I didn't explain my points clearly.

1. First, without `factor` it will fail (this was added 12 years ago by Chris Smith).
Due to `expand_func` it will return `2/(k**2 + 2*k + 1)`.
```python
assert hypersimp(2**k/factorial(k)**2, k) == 2/(k + 1)**2
```
So I cannot remove `factor()`.

3. Here by `poorer` I meant `slower`. (As of I tested, it passed all the test cases). Here is the timing,
(From test cases)
i) Using `gammasimp`,
```python
In [1]: from sympy import hypersimp, factorial, binomial, symbols, rsolve_hyper
   ...: 
   ...: a, n, k = symbols("a n k", integer=True)
   ...: 
   ...: term1 = factorial(k)
   ...: term2 = factorial(k**2)
   ...: term3 = 1/factorial(k)
   ...: term4 = 2**k/factorial(k)**2
   ...: term5 = binomial(n, k)
   ...: term6 = binomial(n + 1, k)
   ...: term7 = (4*k + 1)*factorial(k)/factorial(2*k + 1)
   ...: term8 = 1/((2*k - 1)*factorial(2*k + 1))
   ...: term9 = binomial(n, k)*(-1)**k/factorial(k)
   ...: term10 = binomial(n+1, k)*(2**(-n-1)) - binomial(n, k)*(2**(-n))
   ...: term11 = 2**(-n - 1)*binomial(n + 1, k) - binomial(n, k)/2**n
   ...: term12 = -2**(1 - k)*binomial(k + n - 1, k) + binomial(k + n, k)/2**k
   ...: term13 = -2**(1 - k)*factorial(k + 2)/factorial(k - 1) + factorial(k + 3)/(2**k*factorial(k))
   ...: term14 = 3**k*binomial(k + n, k)/factorial(k)
   ...: term15 = 2**k*binomial(k, 2)/(k + 1)
   ...: term16 = (2*k + 1)*binomial(n, k)
   ...: term17 = (k + 1)*binomial(n, k)/(k + 2)

In [2]: def test1_hyper():
   ...:     hypersimp(term1, k)
   ...:     hypersimp(term2, k)
   ...:     hypersimp(term3, k)
   ...:     hypersimp(term4, k)
   ...:     hypersimp(term5, k)
   ...:     hypersimp(term6, k)
   ...:     hypersimp(term7, k)
   ...:     hypersimp(term8, k)
   ...:     hypersimp(term9, k)
   ...:     hypersimp(term10, k)
   ...:     hypersimp(term11, k)
   ...:     hypersimp(term12, k)
   ...:     hypersimp(term13, k)
   ...:     hypersimp(term14, k)
   ...:     hypersimp(term15, k)
   ...:     hypersimp(term16, k)
   ...:     hypersimp(term17, k)
   ...: 

In [3]: def test1_recurr():
   ...:     rsolve_hyper([-1, -1, 1], 0, n)
   ...:     rsolve_hyper([n**2 - 2, -2*n - 1, 1], 0, n)
   ...:     rsolve_hyper([n**2 - k, -2*n - 1, 1], 0, n)
   ...:     rsolve_hyper([2*n*(n + 1), -n**2 - 3*n + 2, n - 1], 0, n)
   ...:     rsolve_hyper([n + 2, -(2*n + 3)*(17*n**2 + 51*n + 39), n + 1], 0, n)
   ...:     rsolve_hyper([-n - 1, -1, 1], 0, n)
   ...:     rsolve_hyper([-1, 1], n, n)
   ...:     rsolve_hyper([-1, 1], 1 + n, n).expand()
   ...:     rsolve_hyper([-1, 1], 3*(n + n**2), n).expand()
   ...:     rsolve_hyper([-a, 1],0,n)
   ...:     rsolve_hyper([-a, 0, 1], 0, n)
   ...:     rsolve_hyper([1, 1, 1], 0, n).expand()
   ...:     rsolve_hyper([1, -2*n/a - 2/a, 1], 0, n)
   ...: 

In [4]: %timeit test1_hyper()
312 ms ± 31.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [5]: %timeit test1_recurr()
371 ms ± 18.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

ii) Using `collect(expand_func(g.rewrite(gamma))).factor()`
```python
In [1]: from sympy import hypersimp, factorial, binomial, symbols, rsolve_hyper
   ...: 
   ...: a, n, k = symbols("a n k", integer=True)
   ...: 
   ...: term1 = factorial(k)
   ...: term2 = factorial(k**2)
   ...: term3 = 1/factorial(k)
   ...: term4 = 2**k/factorial(k)**2
   ...: term5 = binomial(n, k)
   ...: term6 = binomial(n + 1, k)
   ...: term7 = (4*k + 1)*factorial(k)/factorial(2*k + 1)
   ...: term8 = 1/((2*k - 1)*factorial(2*k + 1))
   ...: term9 = binomial(n, k)*(-1)**k/factorial(k)
   ...: term10 = binomial(n+1, k)*(2**(-n-1)) - binomial(n, k)*(2**(-n))
   ...: term11 = 2**(-n - 1)*binomial(n + 1, k) - binomial(n, k)/2**n
   ...: term12 = -2**(1 - k)*binomial(k + n - 1, k) + binomial(k + n, k)/2**k
   ...: term13 = -2**(1 - k)*factorial(k + 2)/factorial(k - 1) + factorial(k + 3)/(2**k*factorial(k))
   ...: term14 = 3**k*binomial(k + n, k)/factorial(k)
   ...: term15 = 2**k*binomial(k, 2)/(k + 1)
   ...: term16 = (2*k + 1)*binomial(n, k)
   ...: term17 = (k + 1)*binomial(n, k)/(k + 2)

In [2]: def test2_hyper():
   ...:     hypersimp(term1, k)
   ...:     hypersimp(term2, k)
   ...:     hypersimp(term3, k)
   ...:     hypersimp(term4, k)
   ...:     hypersimp(term5, k)
   ...:     hypersimp(term6, k)
   ...:     hypersimp(term7, k)
   ...:     hypersimp(term8, k)
   ...:     hypersimp(term9, k)
   ...:     hypersimp(term10, k)
   ...:     hypersimp(term11, k)
   ...:     hypersimp(term12, k)
   ...:     hypersimp(term13, k)
   ...:     hypersimp(term14, k)
   ...:     hypersimp(term15, k)
   ...:     hypersimp(term16, k)
   ...:     hypersimp(term17, k)
   ...: 

In [3]: def test2_recurr():
   ...:     rsolve_hyper([-1, -1, 1], 0, n)
   ...:     rsolve_hyper([n**2 - 2, -2*n - 1, 1], 0, n)
   ...:     rsolve_hyper([n**2 - k, -2*n - 1, 1], 0, n)
   ...:     rsolve_hyper([2*n*(n + 1), -n**2 - 3*n + 2, n - 1], 0, n)
   ...:     rsolve_hyper([n + 2, -(2*n + 3)*(17*n**2 + 51*n + 39), n + 1], 0, n)
   ...:     rsolve_hyper([-n - 1, -1, 1], 0, n)
   ...:     rsolve_hyper([-1, 1], n, n)
   ...:     rsolve_hyper([-1, 1], 1 + n, n).expand()
   ...:     rsolve_hyper([-1, 1], 3*(n + n**2), n).expand()
   ...:     rsolve_hyper([-a, 1],0,n)
   ...:     rsolve_hyper([-a, 0, 1], 0, n)
   ...:     rsolve_hyper([1, 1, 1], 0, n).expand()
   ...:     rsolve_hyper([1, -2*n/a - 2/a, 1], 0, n)
   ...: 

In [4]: %timeit test2_hyper()
337 ms ± 23.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [5]: %timeit test2_recurr()
397 ms ± 59.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

4. `gammasimp` does call `factor` (not `cancel`), but after a slight simplification only. Still, it improves the time slightly atleast.

When the respective functions ran for 10 times,
Case i) `gammasimp`,
```python
In [8]: clear_cache()

In [9]: for i in range(10):
   ...:     clear_cache()
   ...:     %timeit test1_hyper()
   ...: 
308 ms ± 15.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
303 ms ± 18.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
305 ms ± 36.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
310 ms ± 31.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
307 ms ± 32.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
314 ms ± 43.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
332 ms ± 51.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
306 ms ± 38.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
298 ms ± 19.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
291 ms ± 8.18 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [10]: for i in range(10):
    ...:     clear_cache()
    ...:     %timeit test1_recurr()
    ...: 
381 ms ± 44.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
387 ms ± 23.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
382 ms ± 33 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
395 ms ± 32.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
382 ms ± 37.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
367 ms ± 13.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
374 ms ± 24.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
378 ms ± 32.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
405 ms ± 40.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
439 ms ± 127 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
Similarly for case ii) `cancel(expand_func(g.rewrite(gamma))).factor()`,
```python
In [4]: for i in range(10):
   ...:     clear_cache()
   ...:     %timeit test2_recurr()
   ...: 
382 ms ± 28.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
374 ms ± 33.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
377 ms ± 27.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
483 ms ± 114 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
382 ms ± 42.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
374 ms ± 23.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
366 ms ± 18.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
386 ms ± 32.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
362 ms ± 13.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
361 ms ± 19 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [5]: for i in range(10):
   ...:     clear_cache()
   ...:     %timeit test2_hyper()
   ...: 
335 ms ± 20.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
364 ms ± 27.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
333 ms ± 9.31 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
347 ms ± 28.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
342 ms ± 35.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
338 ms ± 23.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
339 ms ± 35.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
331 ms ± 21.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
368 ms ± 38.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
341 ms ± 12 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

So, `gammasimp` is somehow slightly faster than its alternative.

oscarbenjamin commented:
> 1. First, without `factor` it will fail (this was added 12 years ago by Chris Smith).
>    Due to `expand_func` it will return `2/(k**2 + 2*k + 1)`.
> 
> ```python
> assert hypersimp(2**k/factorial(k)**2, k) == 2/(k + 1)**2
> ```

Those answers are equivalent. The test could just be changed.

Are there any other cases that need `factor`?

How do the timings compare if you only use `cancel` and not `factor` or what if you only use `factor` and not `cancel`?

In general `timeit` is not a good measure with SymPy because of the cache. Clearing the cache before calling timeit doesn't really work to get a good timing because timeit itself uses a loop and you really need the cache to be cleared in between each iteration of that loop.

The best way to time something like this that makes use of caching is just to restart Python and use `%time` but run some other similar command first to make sure everything is loaded e.g. we can time this with master:
```python
In [1]: rsolve_hyper([-1, -1, 1], 0, n)
Out[1]: 
           n              n
   ⎛1   √5⎞       ⎛1   √5⎞ 
C₀⋅⎜─ - ──⎟  + C₁⋅⎜─ + ──⎟ 
   ⎝2   2 ⎠       ⎝2   2 ⎠ 

In [2]: %time rsolve_hyper([n**2 - 2, -2*n - 1, 1], 0, n)
CPU times: user 179 ms, sys: 5 μs, total: 179 ms
Wall time: 178 ms
Out[2]: C₀⋅RisingFactorial(-√2, n) + C₁⋅RisingFactorial(√2, n)
```
This is with `gammasimp`:
```python
In [2]: %time rsolve_hyper([n**2 - 2, -2*n - 1, 1], 0, n)
CPU times: user 212 ms, sys: 4 ms, total: 216 ms
Wall time: 216 ms
Out[2]: C₀⋅RisingFactorial(-√2, n) + C₁⋅RisingFactorial(√2, n)
```
This is with `cancel(expand_func(g.rewrite(gamma)))`:
```python
In [2]: %time rsolve_hyper([n**2 - 2, -2*n - 1, 1], 0, n)
CPU times: user 172 ms, sys: 3.97 ms, total: 176 ms
Wall time: 175 ms
Out[2]: C₀⋅RisingFactorial(-√2, n) + C₁⋅RisingFactorial(√2, n)
```
This is with `factor(expand_func(g.rewrite(gamma)))`:
```python
In [2]: %time rsolve_hyper([n**2 - 2, -2*n - 1, 1], 0, n)
CPU times: user 163 ms, sys: 0 ns, total: 163 ms
Wall time: 162 ms
Out[2]: C₀⋅RisingFactorial(-√2, n) + C₁⋅RisingFactorial(√2, n)
```
At least for these measurements `gammasimp` is the slowest. The other three methods are all about the same time considering the potential variation from repeating the experiment.

The reason that I suggest `cancel(expand_func(g.rewrite(gamma)))` rather than `gammasimp` is because generally `cancel` is the standard way to put an expression into canonical form and I know that we do need what each of these three steps is doing:

- We need rewrite gamma to turn factorials, binomial etc into gamma.
- We need expand_func to handle things like `gamma(k + 1) / gamma(k)`.
- We need cancel to cancel out non-rational factors like `2**k` and `gamma(k)` after `expand_func`.

I don't know exactly what are all the things that `gammasimp` does and it is not clear to me that they are all particularly helpful in the case of `hypersimp`.

Harikrishna-Srinivasan commented:
My conclusion is there are 3 diff. ways for this.

(This is one of ~10 tests, that has been done by restarting `bin/isympy` each time, so there is no talk about `caching`)

1) `gammasimp(g)`:
```python
In [2]: %time test1_hyper()
CPU times: user 533 ms, sys: 6.48 ms, total: 539 ms
Wall time: 549 ms

In [3]: %time test1_recurr()
CPU times: user 602 ms, sys: 4 ms, total: 606 ms
Wall time: 610 ms
```
2) `cancel(expand_func(g.rewrite(gamma)))` - Fails few tests that can be changed:
```python
In [2]: %time test1_hyper()
CPU times: user 642 ms, sys: 4.96 ms, total: 647 ms
Wall time: 648 ms

In [3]: %time test1_recurr()
CPU times: user 571 ms, sys: 948 μs, total: 572 ms
Wall time: 571 ms
```

3) `factor(expand_func(g.rewrite(gamma)))`:
```python
In [2]: %time test1_hyper()
CPU times: user 526 ms, sys: 5.59 ms, total: 531 ms
Wall time: 537 ms

In [3]: %time test1_recurr()
CPU times: user 659 ms, sys: 2.93 ms, total: 662 ms
Wall time: 708 ms
```

Observations: Speed-wise 2 > 3 > 1

I'm changing to 2) `cancel(expand_func(g.rewrite(gamma)))` & the test cases that can be equivalently expressed. (without factoring)

Harikrishna-Srinivasan commented:
Actually, now I found out that `cancel(expand_func(g.rewrite(gamma)))` fails in few test cases.

Like `hypersimp(n*(n + a + b)*a**n*b**n/(factorial(n + a)*factorial(n + b)))` returned None.
But expected is `a*b*(n + 1)*(a + b + n + 1)/(n*(a + b + n)*(a + n + 1)*(b + n + 1))` in one of the `gosper_sum` test.
\# line 54 in test_gosper.py (+ many other tests too)
```python
...
    # issue 6033:
    assert gosper_sum(
        n*(n + a + b)*a**n*b**n/(factorial(n + a)*factorial(n + b)), \
        (n, 0, m)).simplify() == -exp(m*log(a) + m*log(b))*gamma(a + 1) \
        *gamma(b + 1)/(gamma(a)*gamma(b)*gamma(a + m + 1)*gamma(b + m + 1)) \
        + 1/(gamma(a)*gamma(b))
```
So I'm reverting everything back & add the failed tests in `test_simplify.py` also.

Harikrishna-Srinivasan committed:
Add tests and revert to gammasimp

Harikrishna-Srinivasan committed:
Corrected simplify tests

oscarbenjamin commented:
> Like `hypersimp(n*(n + a + b)*a**n*b**n/(factorial(n + a)*factorial(n + b)))` returned None.
> 
> But expected is `a*b*(n + 1)*(a + b + n + 1)/(n*(a + b + n)*(a + n + 1)*(b + n + 1))` in one of the `gosper_sum` test.

Okay but now the question is why? Why does `gammasimp` make this work if the other options do not?

What would be the best way to make this work?

What are the features of different expressions that hypersimp might be expected to handle?

This example has `a**n` and `b**n`. That means that the term ratio has e.g. `a**(n+1)/a**n` which should ideally simplify to `a`. Calling cancel before powsimp prevents that from simplifying but calling `cancel` after does not.

Thinking about this now I think it should be possible to separate the factors of `f` and consider the ratio for each in turn e.g.:
```python
In [16]: a, b, n = symbols('a, b, n')

In [17]: f = n*(n + a + b)*a**n*b**n/(factorial(n + a)*factorial(n + b))

In [18]: f
Out[18]: 
 n  n              
a ⋅b ⋅n⋅(a + b + n)
───────────────────
 (a + n)!⋅(b + n)! 

In [19]: Mul.make_args(factor_terms(f.rewrite(gamma)))
Out[19]: 
⎛    n   n       1             1                 ⎞
⎜n, a , b , ────────────, ────────────, a + b + n⎟
⎝           Γ(a + n + 1)  Γ(b + n + 1)           ⎠
```
If these are `e_i(n)` then our goal is to identify when `g_i(n) = e_i(n+1)/e_i(n)` will be a rational function of `n`. If `e_i(n)` is a rational function of `n` then `g_i(n)` will be as well. If `e_i(n)` is of the form `gamma(n + e)` then `g_i(n)` will be a rational function of `n`. If `e_i(n)` is of the form `e1**(n+e2)` then it will be. We can at least extract simple factors like this to get the term ratio before considering anything like `gammasimp`.

oscarbenjamin commented:
Maybe this is better:
```python
In [1]: a, b, n = symbols('a, b, n')

In [2]: f = n*(n + a + b)*a**n*b**n/(factorial(n + a)*factorial(n + b))

In [3]: es = Mul.make_args(factor_terms(expand_func(f.rewrite(gamma))))

In [4]: es
Out[4]: 
⎛    n   n    1      1       1         1               ⎞
⎜n, a , b , ─────, ─────, ────────, ────────, a + b + n⎟
⎝           a + n  b + n  Γ(a + n)  Γ(b + n)           ⎠

In [5]: [e.subs(n, n+1) / e for e in es]
Out[5]: 
⎡n + 1   -n  n + 1   -n  n + 1    a + n      b + n      Γ(a + n)      Γ(b + n)    a + b + n + 1⎤
⎢─────, a  ⋅a     , b  ⋅b     , ─────────, ─────────, ────────────, ────────────, ─────────────⎥
⎣  n                            a + n + 1  b + n + 1  Γ(a + n + 1)  Γ(b + n + 1)    a + b + n  ⎦

In [6]: [powsimp(e.subs(n, n+1) / e) for e in es]
Out[6]: 
⎡n + 1          a + n      b + n      Γ(a + n)      Γ(b + n)    a + b + n + 1⎤
⎢─────, a, b, ─────────, ─────────, ────────────, ────────────, ─────────────⎥
⎣  n          a + n + 1  b + n + 1  Γ(a + n + 1)  Γ(b + n + 1)    a + b + n  ⎦

In [7]: [powsimp(expand_func(e.subs(n, n+1) / e)) for e in es]
Out[7]: 
⎡n + 1          a + n      b + n      1      1    a + b + n + 1⎤
⎢─────, a, b, ─────────, ─────────, ─────, ─────, ─────────────⎥
⎣  n          a + n + 1  b + n + 1  a + n  b + n    a + b + n  ⎦

In [10]: Mul(*[powsimp(expand_func(e.subs(n, n+1) / e), combine='exp') for e in es])
Out[10]: 
     a⋅b⋅(n + 1)⋅(a + b + n + 1)     
─────────────────────────────────────
n⋅(a + b + n)⋅(a + n + 1)⋅(b + n + 1)
```

Harikrishna-Srinivasan commented:
This arises one more problem.

That is when symbols are non-commutative, `Mul` may not obey that & raises `sympy.polys.polyerrors.PolynomialError: non-commutative expressions are not supported`

So, we have to look for some other option.
`gammasimp` obeys non-commutativity correctly (still we can do some alterations & make it work)

oscarbenjamin commented:
> That is when symbols are non-commutative, `Mul` may not obey that & raises `sympy.polys.polyerrors.PolynomialError: non-commutative expressions are not supported`

Once again can you give an example?

I don't see how `hypersimp` can do anything useful with non-commutative expressions.

Harikrishna-Srinivasan commented:
For
```python
from sympy.abc import *
from sympy import *

A, B = symbols("A B", commutative=False)
M = symbols('M', integer=True, positive=True)
print(hypersimp(A**k * B**k, k))
```
Expected: None
Output: A*B

Because these are non-commutatives, we can't directly apply this (but needs slight modifications).

I got the error for (from test case `test_sums_products.py` in `test_noncommutativity_honoured()`)
```python
p = Sum(B**n*A*B**n, (n, 1, M))
print(summation(p, (n, 1, M)))
```

oscarbenjamin commented:
Can you show what it is that raises `PolynomialError: non-commutative expressions are not supported`. I don't think that Mul will raise that error.

Generally hypersimp cannot do anything with non-commutative expressions because it simplifies the term ratio but that is not something that can be defined for non-commutative expressions.

oscarbenjamin commented:
The traceback you showed doesn't involve `hypersimp`.

Harikrishna-Srinivasan commented:
See https://github.com/sympy/sympy/pull/27931#issuecomment-2799939376, due to this wrong output, this error has occurred

oscarbenjamin commented:
> See [#27931 (comment)](https://github.com/sympy/sympy/pull/27931#issuecomment-2799939376), due to this wrong output, this error has occurred

Please think more carefully about how you communicate and how anyone is supposed to understand what you are saying. You are referring to this:

> ```python
> from sympy.abc import *
> from sympy import *
> 
> A, B = symbols("A B", commutative=False)
> M = symbols('M', integer=True, positive=True)
> print(hypersimp(A**k * B**k, k))
> ```
> Expected: None
> Output: A*B

This is referring to code that you have not shown. Somehow hypersimp gives a wrong output but it is not the hypersimp that is on the master branch and it is not the hypersimp that is in the PR. You made some unspecified changes to hypersimp and now you see this but I have no idea what you changed. Instead of showing the output from hypersimp show what actual code that you used in hypersimp produced this.

I guess you mean something like this:
```python
In [3]: A, B = symbols("A B", commutative=False)
   ...: M = symbols('M', integer=True, positive=True)

In [4]: f = A**k*B**k

In [5]: es = Mul.make_args(factor_terms(expand_func(f.rewrite(gamma))))

In [6]: es
Out[6]: 
⎛ k   k⎞
⎝A , B ⎠

In [8]: Mul(*[powsimp(expand_func(e.subs(k, k+1) / e), combine='exp') for e in es])
Out[8]: A⋅B
```
The simple fix here is just that `hypersimp` should return `None` is the expression is not commutative. It could perhaps use `args_cnc` rather than `Mul.make_args` and if there are any non-commutative args it can return `None`.

Harikrishna-Srinivasan commented:
Okay, I'll try to be more clearer hereafter. 
I was actually mentioning, 
```python
In [3]: A, B = symbols("A B", commutative=False)
   ...: M = symbols('M', integer=True, positive=True)

In [4]: f = A**k*B**k

In [5]: es = Mul.make_args(factor_terms(expand_func(f.rewrite(gamma))))

In [6]: es
Out[6]: 
⎛ k   k⎞
⎝A , B ⎠

In [8]: Mul(*[powsimp(expand_func(e.subs(k, k+1) / e), combine='exp') for e in es])
Out[8]: A⋅B
```
only.
And I'll try with `args_cnc` (I didn't know this earlier) instead of `Mul`.

Harikrishna-Srinivasan commented:
I was foolish for not looking into some of the details in `gammasimp`.

`factor()` function is enough for us to pass all tests. But we should not factor the terms in 2 cases,

1) When there is no `gamma` function.
2) When there is an `exp` and it has arguments that contain `k` (the parameter).

The change will be like,
```python
def hypersimp(f, k):
    """
    No changes done in docs.
    """

    f = sympify(f)
    g = f.subs(k, k+1) / f
    g = g.rewrite(gamma)

    if g.is_commutative is False:
        return None

    has_gamma = False
    has_exp = False
    for expr in g.atoms(Function):  # checking for `gamma` & `exp`
        if expr.func == gamma:
            has_gamma = True
        elif expr.func == exp and expr.has(k): # line (18)
            has_exp = True
        if has_gamma and has_exp:
            break

    if g.has(Piecewise):
        g = piecewise_fold(g)
        g = g.args[-1][0]

    g = expand_func(g)
    g = powsimp(g, deep=True, combine="exp")

    if has_gamma and has_exp is False:
        g = factor(g) # factor `g` if it has `gamma` func. & not an `exp` with `k`

    if g.is_rational_function(k):
        return simplify(g, ratio=S.Infinity)
    else:
        return None
```
It is still not clear whether I should or not, Omit `expr.has(k)` in **Line 18**. It works even if it is not present, but would simplify more if it is there. (It passes all of the related test cases, not sure about other tests that might depend on `hypersimp`, yet)

Harikrishna-Srinivasan committed:
gammasimp->factor for perf.

oscarbenjamin commented:
> But we should not factor the terms in 2 cases,
> 
> 1. When there is no `gamma` function.
> 2. When there is an `exp` and it has arguments that contains `k` (the parameter).

Why not?

In any case I think I see now that this should use a different strategy. The problem is in the beginning with `g = f.subs(k, k+1) / f`. It is better to first split `f` and then figure out how the ratio of each factor in `f` can be written as a rational function. Then I am that not sure that factor is needed at all.

Harikrishna-Srinivasan commented:
> > But we should not factor the terms in 2 cases,
> > 
> > 1. When there is no `gamma` function.
> > 2. When there is an `exp` and it has arguments that contains `k` (the parameter).
> 
> Why not?

1) Why not factor() if it has no `gamma` fn: (_fn_ short for **function**)

We rewrite `g` in terms of `gamma` & if it is not present no more simplification is needed.
_i.e.,_ given expression needs no further `factor` to tell whether `g` is rational_function of `k` or not. (it is in its simplest form, rational form)

2) If it contains `exp` with `k` as an argument, then `factor` will cancel these terms. Here is an example,
(Code as in https://github.com/sympy/sympy/pull/27931#issuecomment-2811508430)
```python
from sympy.abc import *
from sympy import *

f = exp(k*t)*gamma(k + t)/factorial(k)
# g = exp(-k*t)*exp(t*(k + 1))*factorial(k)*gamma(k + t + 1)/(factorial(k + 1)*gamma(k + t))
print(hypersimp(f, k))
```
Output: None
\# if we had used `factor` for this case, we would have obtained `(k + t)*exp(t)/(k + 1)`, which is a rational fn in `k`.
(`f` is actually hypergeometric only)

The reason is not mathematical. I saw that this test case expects `hypersimp` to return *None*.
(In `stats/tests/test_compund_rv.py` near line 47)
```python
def test_poisson_CompoundDist():
    k, t, y = symbols('k t y', positive=True, real=True)
    G = Gamma('G', k, t)
    D = Poisson('P', G)
    assert density(D)(y).simplify() == t**y*(t + 1)**(-k - y)*gamma(k + y)/(gamma(k)*gamma(y + 1))
    # https://en.wikipedia.org/wiki/Negative_binomial_distribution#Gamma%E2%80%93Poisson_mixture
    assert E(D).simplify() == k*t # mean of NegativeBinomialDistribution
```
I get,
```python
____________________________________________________________________ test_poisson_CompoundDist ____________________________________________________________________

self = Piecewise((gamma(k)/(-t*exp(_x)/(t + 1) + 1)**k, ((k >= 1) & (t*exp(_x)/(t + 1) < 1)) | ((k < 1) & (t*exp(_x)/(t + 1) ...))), (Sum(t**_x*exp(_x*_x)*gamma(_x + k)/((t + 1)**_x*factorial(_x)), (_x, 0, oo)), True))/(t**k*(1 + 1/t)**k*gamma(k))
x = _x, n = 2, logx = None, cdir = 1

    def _eval_nseries(self, x, n, logx, cdir=0):
        from .function import PoleError
        from sympy.functions.elementary.integers import ceiling
        from sympy.series.order import Order
    
        def coeff_exp(term, x):
            lt = term.as_coeff_exponent(x)
            if lt[0].has(x):
                try:
                    lt = term.leadterm(x)
                except ValueError:
                    return term, S.Zero
            return lt
    
        ords = []
    
        try:
            for t in self.args:
                coeff, exp = t.leadterm(x)
                if not coeff.has(x):
                    ords.append((t, exp))
                else:
                    raise ValueError
    
            n0 = sum(t[1] for t in ords if t[1].is_number)
            facs = []
            for t, m in ords:
                n1 = ceiling(n - n0 + (m if m.is_number else 0))
>               s = t.nseries(x, n=n1, logx=logx, cdir=cdir)

sympy/core/mul.py:1990: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
sympy/core/expr.py:3466: in nseries
    return self._eval_nseries(x, n=n, logx=logx, cdir=cdir)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = Piecewise((gamma(k)/(-t*exp(_x)/(t + 1) + 1)**k, ((k >= 1) & (t*exp(_x)/(t + 1) < 1)) | ((k < 1) & (t*exp(_x)/(t + 1) <= 1) & Ne(t*exp(_x)/(t + 1), 1))), (Sum(t**_x*exp(_x*_x)*gamma(_x + k)/((t + 1)**_x*factorial(_x)), (_x, 0, oo)), True))
x = _x, n = 2, logx = None, cdir = 1

    def _eval_nseries(self, x, n, logx, cdir=0):
>       args = [(ec.expr._eval_nseries(x, n, logx), ec.cond) for ec in self.args]
E       TypeError: Expr._eval_nseries() missing 1 required positional argument: 'cdir'

sympy/functions/elementary/piecewise.py:751: TypeError

During handling of the above exception, another exception occurred:

    def test_poisson_CompoundDist():
        k, t, y = symbols('k t y', positive=True, real=True)
        G = Gamma('G', k, t)
        D = Poisson('P', G)
        assert density(D)(y).simplify() == t**y*(t + 1)**(-k - y)*gamma(k + y)/(gamma(k)*gamma(y + 1))
        # https://en.wikipedia.org/wiki/Negative_binomial_distribution#Gamma%E2%80%93Poisson_mixture
>       assert E(D).simplify() == k*t # mean of NegativeBinomialDistribution

sympy/stats/tests/test_compound_rv.py:53: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
sympy/stats/rv.py:808: in expectation
    return Expectation(expr, condition).doit(**kwargs)
sympy/stats/symbolic_probability.py:288: in doit
    result = pspace(expr).compute_expectation(expr, evaluate=evaluate)
sympy/stats/compound_rv.py:114: in compute_expectation
    return new_pspace.compute_expectation(expr, rvs, evaluate, **kwargs)
sympy/stats/drv.py:313: in compute_expectation
    return self.distribution.expectation(expr, x, evaluate=evaluate,
sympy/stats/drv.py:162: in expectation
    taylor = poly(series(mgf, t, 0, deg + 1).removeO(), t)
sympy/series/series.py:63: in series
    return expr.series(x, x0, n, dir)
sympy/core/expr.py:3084: in series
    rv = self.subs(x, xpos).series(xpos, x0, n, dir, logx=logx, cdir=cdir)
sympy/core/expr.py:3092: in series
    s1 = self._eval_nseries(x, n=n, logx=logx, cdir=cdir)
sympy/core/mul.py:2004: in _eval_nseries
    facs = [t.nseries(x, n=ceiling(n-n0), logx=logx, cdir=cdir) for t in self.args]
sympy/core/expr.py:3466: in nseries
    return self._eval_nseries(x, n=n, logx=logx, cdir=cdir)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = Piecewise((gamma(k)/(-t*exp(_x)/(t + 1) + 1)**k, ((k >= 1) & (t*exp(_x)/(t + 1) < 1)) | ((k < 1) & (t*exp(_x)/(t + 1) <= 1) & Ne(t*exp(_x)/(t + 1), 1))), (Sum(t**_x*exp(_x*_x)*gamma(_x + k)/((t + 1)**_x*factorial(_x)), (_x, 0, oo)), True))
x = _x, n = 2, logx = None, cdir = 1

    def _eval_nseries(self, x, n, logx, cdir=0):
>       args = [(ec.expr._eval_nseries(x, n, logx), ec.cond) for ec in self.args]
E       TypeError: Expr._eval_nseries() missing 1 required positional argument: 'cdir'

sympy/functions/elementary/piecewise.py:751: TypeError
                                                                         DO *NOT* COMMIT!                                                                         
===================================================================== short test summary info =====================================================================
FAILED sympy/stats/tests/test_compound_rv.py::test_poisson_CompoundDist - TypeError: Expr._eval_nseries() missing 1 required positional argument: 'cdir'
```
> [!NOTE]
> This is also due to `hypersimp` only indirectly. (only if we factor out when `exp` term with `k` as argument is present. *i.e.,* case 2)
I'm not sure if this is a probl with `_eval_nseries` or any other related fn (or no problem at all).

> In any case I think I see now that this should use a different strategy. The problem is in the beginning with `g = f.subs(k, k+1) / f`. It is better to first split `f` and then figure out how the ratio of each factor in `f` can be written as a rational function. Then I am that not sure that factor is needed at all.

I don't quite understand, why you ask me to use a different approach like `Mul` & `.args_cnc()`, if `factor()` does this quite easily. I will get an idea if you tell me why a different strategy is needed.

oscarbenjamin commented:
> I don't quite understand, why you ask me to use a different approach like `Mul` & `.args_cnc()`, if `factor()` does this quite easily.

The factor function can radically transform an expression and can be very expensive for some expressions. It is not clear to me that there are any situations where it is actually needed that would not be handled by my suggestion with `factor_terms` which is much lighter weight.

Harikrishna-Srinivasan commented:
Oh, in that case `factor_terms` can be replace in place of `factor` in this place without any issue.

I thought about it and tested whether `and expr.has(k)` needed, & turns out that it is not (from https://github.com/sympy/sympy/pull/27931#issuecomment-2811508430).

So shall I make the changes?
I'm asking this because I don't see a big speed difference b/w `factor()` & `factor_terms()` in so should I think of alternative approach that is faster, or is this sufficient?

oscarbenjamin commented:
> I don't see a big speed difference b/w `factor()` & `factor_terms()`

Try with higher degree inputs:
```python
In [1]: %time factor(x**1000 + 1)
CPU times: user 707 ms, sys: 11.6 ms, total: 718 ms
Wall time: 715 ms
Out[1]: 
⎛ 8    ⎞ ⎛ 32    24    16    8    ⎞ ⎛ 160    120    80    40    ⎞ ⎛ 800    600    400    200    ⎞
⎝x  + 1⎠⋅⎝x   - x   + x   - x  + 1⎠⋅⎝x    - x    + x   - x   + 1⎠⋅⎝x    - x    + x    - x    + 1⎠

In [2]: %time factor_terms(x**1000 + 1)
CPU times: user 1.73 ms, sys: 93 μs, total: 1.82 ms
Wall time: 1.84 ms
Out[2]: 
 1000    
x     + 1
```
These two functions don't do the same thing but it is possible that `factor_terms` does everything that is needed here. In particular the main thing that `factor` does is to factor a polynomial into irreducible factors but here we don't care about that. We are just simplifying to see if what is left over at the end is a rational function.

> So shall I make the changes?

I would like to see whether it can be made to work by splitting `f` as I described above. As I said before I think the problem is with doing `g = f.subs(k, k+1) / f` and then trying to simplify `g`. This is creating an expression that then has to be simplified but I think we can avoid creating that expression in the first place.

oscarbenjamin commented:
I don't think it is necessary to use `.subs(k, k+1)` at all or at least none of the examples that you have shown have needed it. Please look carefully at this https://github.com/sympy/sympy/pull/27931#issuecomment-2799895224.

If `f` has a factor say `x**k` then we don't need `.subs` to say that the term ratio will be `x`. If it has a factor `gamma(k)`  then the term ratio will be `k` and again we don't need `.subs` to figure that out. If the factor is already a polynomial in `k` then we can do `e.subs(k, k+1) / e` but we don't need to simplify that because we know it is a rational function. If a factor is a power with integer exponent then we don't need subs because we can just compute the term ratio of the base.

I don't see yet why subs or simplification functions are needed because every example is something where the term ratio can be calculated directly.

Harikrishna-Srinivasan commented:
I think I got your point, so I changed the code to

```python
def hypersimp(f, k):
    def hyper(f):
        fns = {gamma, exp, Pow, Piecewise}
        def is_rational_ratio(expr):
            if expr.is_rational_function(k):
                return True
            if expr.func not in fns:
                return False
            args = expr.args
            if expr.func == Piecewise:
                return is_hyper(args[-1][0])
            if expr.func == gamma:
                return args[0].diff(k, 2).is_zero
            if expr.func == exp:
                return args[0].diff(k, 1).is_Number
            if expr.func == Pow:
                if args[0] is k:
                    return args[1].is_rational
                else:
                    return args[1].diff(k).is_rational
            return True

        def is_hyper(f):
            f = f.rewrite(gamma)
            args = Add.make_args(f)
            return all(
                all(map(is_rational_ratio, Mul.make_args(numer) + Mul.make_args(den)))
                for arg in args
                for numer, den in {fraction(arg)}
            )

        if f.is_rational_function(k):
            return f.subs(k, k + 1) / f

        if is_hyper(f):
            return f.subs(k, k + 1) / f
        return None


    f = sympify(f)
    f = f.rewrite(gamma)

    if not f.is_commutative:
        return None

    g = hyper(f)
    if g is None:
        return None

    g = (f.subs(k, k + 1)/f).rewrite(gamma)
    g = expand_func(g)
    g = powsimp(g, deep=True, combine="exp")
    if g.has(Piecewise):
        g = piecewise_fold(g)
        g = g.args[-1][0]

    return simplify(g, ratio=S.Infinity)
```
I think it would be better to have `hyper(f)` as `is_hypergeometric(self, k)` in `Basic` class in `basic.py`. (In my code I assumed: For `Piecewise` I'll only check its last arg., but `is_hypergeometric` returns None if expr is `Piecewise`).

Did I get your point right?

Harikrishna-Srinivasan committed:
Merge branch 'sympy:master' into master

oscarbenjamin commented:
That's sort of what I meant but it can be improved a lot. Rather than having a `is_hyper` function there should be a function called e.g. `get_rational_term_ratio` that either returns the term ratio as a rational function or returns None. Then you can do `Mul.make_args(factor_terms(expr))` and call `get_rational_term_ratio` on each factor.

By the way why does this do:
```python
    if g.has(Piecewise):
        g = piecewise_fold(g)
        g = g.args[-1][0]
```
I know you didn't add that but it is definitely wrong. You can't just pick an arbitrary case from a Piecewise like that. It isn't even guaranteed that `piecewise_fold` would return a `Piecewise` e.g.:
```python
In [10]: p1 = Piecewise((x, (x>=0)))

In [11]: p2 = Piecewise((p1, x>0), (x, True))

In [12]: p2
Out[12]: 
⎧{x  for x ≥ 0  for x > 0
⎨                        
⎩      x        otherwise

In [13]: piecewise_fold(p2)
Out[13]: x

In [14]: hypersimp(p2, x)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[14], line 1
----> 1 hypersimp(p2, x)

File /.../sympy/simplify/simplify.py:323, in hypersimp(f, k)
    321 if g.has(Piecewise):
    322     g = piecewise_fold(g)
--> 323     g = g.args[-1][0]
    324 g = expand_func(g)
    325 g = powsimp(g, deep=True, combine='exp')

TypeError: 'Add' object is not subscriptable
```
It is potentially reasonable to call `piecewise_fold` but it is not reasonable to try to pick a case from the Piecewise.

Do any tests fail if that part is removed?

Harikrishna-Srinivasan commented:
Yes, it will fails some of the tests if
```python
    if g.has(Piecewise):
        g = piecewise_fold(g)
        g = g.args[-1][0]
```
is removed. Like,
```python
def test_product_spaces():
    X1 = Geometric('X1', S.Half)
    X2 = Geometric('X2', Rational(1, 3))
    assert str(P(X1 + X2 < 3).rewrite(Sum)) == (
        "Sum(Piecewise((1/(4*2**n), n >= -1), (0, True)), (n, -oo, -1))/3")
    assert str(P(X1 + X2 > 3).rewrite(Sum)) == (
        'Sum(Piecewise((2**(X2 - n - 2)*(2/3)**(X2 - 1)/6, '
        'X2 - n <= 2), (0, True)), (X2, 1, oo), (n, 1, oo))')
    assert P(Eq(X1 + X2, 3)) == Rational(1, 12)
```
```
FAILED sympy/stats/tests/test_discrete_rv.py::test_product_spaces - sympy.polys.polyerrors.PolynomialError: Piecewise((4/3, (_d <= _d) & (_d - _d <= 1)), (2**_d*zoo, _d <= _d), (0, _d - _d <= 1), (nan, True))
```
Just to be clear, I should calculate the ratio (without using `.subs(k, k+1)`) & then check if its rational function in `k`, right?

oscarbenjamin commented on sympy/solvers/recurr.py:
```diff
@@ -550,6 +550,9 @@ def rsolve_hyper(coeffs, f, n, **hints):
             s = hypersimp(g, n)
 
+            if s is None:
+                return None
```
Just so we are clear here. This is the actual fix for bug in the cited issue.

I would be happy to merge a PR that only has this fix along with a regression test for the case in the issue.

Improving hypersimp is a separate more complicated question which is why this is turning into a lengthier discussion.

Harikrishna-Srinivasan replied:
Okay, I'll keep only this change along with its test cases.

oscarbenjamin commented:
> Just to be clear, should I calculate the ratio (without using `.subs(k, k+1)`) & then check if it's a rational function in `k`, Or should I check if the ratio is rational & then calculate the ratio one-by-one for each term?

I feel like I have already explained this above or that this question is something that you can figure out if you just think about it a bit. Can you please read carefully what has already been said and then think for yourself about how to write the code?

I don't actually want to go back and forth like this explaining things again and again until I basically have to write the code myself.

oscarbenjamin commented:
> Yes, it will fail some of the tests if

I just tried removing the Piecewise check and that test passes:
```diff
diff --git a/sympy/simplify/simplify.py b/sympy/simplify/simplify.py
index 8b315cc20c..0c73a8453d 100644
--- a/sympy/simplify/simplify.py
+++ b/sympy/simplify/simplify.py
@@ -318,9 +318,6 @@ def hypersimp(f, k):
     g = f.subs(k, k + 1) / f
 
     g = g.rewrite(gamma)
-    if g.has(Piecewise):
-        g = piecewise_fold(g)
-        g = g.args[-1][0]
     g = expand_func(g)
     g = powsimp(g, deep=True, combine='exp')
 
```
Then:
```python
In [1]: from sympy.stats import *

In [2]:     X1 = Geometric('X1', S.Half)
   ...:     X2 = Geometric('X2', Rational(1, 3))
   ...:     assert str(P(X1 + X2 < 3).rewrite(Sum)) == (
   ...:         "Sum(Piecewise((1/(4*2**n), n >= -1), (0, True)), (n, -oo, -1))/3")
   ...:     assert str(P(X1 + X2 > 3).rewrite(Sum)) == (
   ...:         'Sum(Piecewise((2**(X2 - n - 2)*(2/3)**(X2 - 1)/6, '
   ...:         'X2 - n <= 2), (0, True)), (X2, 1, oo), (n, 1, oo))')
   ...:     assert P(Eq(X1 + X2, 3)) == Rational(1, 12)

In [3]: P(Eq(X1 + X2, 3))
Out[3]: 1/12
```

Harikrishna-Srinivasan committed:
Fix: AttributeError for rsolve_hyper from hypersimp

Harikrishna-Srinivasan committed:
Merge branch 'master' of https://github.com/Harikrishna-Srinivasan/sympy

Harikrishna-Srinivasan committed:
Merge with sympy

Harikrishna-Srinivasan committed:
Corrected back hypersimp tests

Harikrishna-Srinivasan committed:
Uncommit unknown commits

Harikrishna-Srinivasan committed:
Added necessary imports

oscarbenjamin commented on sympy/solvers/tests/test_recurr.py:
```diff
@@ -75,6 +76,18 @@ def test_rsolve_hyper():
 
     assert rsolve_hyper([1, -2*n/a - 2/a, 1], 0, n) == 0
 
+    # issue 27901
+    assert rsolve_hyper([-2*2**n, 2*2**n], -2*binomial(n, k) + binomial(n + 1, k), k) is None
+    assert rsolve_hyper([-1, 1], factorial(n**2), n) is None
+    assert rsolve_hyper([-1, 1], n**n, n) is None
+    assert rsolve_hyper([-1, 1], sqrt(n), n) is None
+
+    assert rsolve_hyper([-2, 1], 2**(n**3 + 1), n) is None
+    assert rsolve_hyper([-2, 1], 2**(a*n + 1), n) is None
+
+    assert rsolve_hyper([-1, 1], fibonacci(n), n) is None
+    assert rsolve_hyper([-1, 1], euler(n), n) is None
+    assert rsolve_hyper([-1, 1], harmonic(n), n) is None
```
Are these all examples that should really not be None if hypersimp was improved? There should at least be some explanation in comments about this.

Use a URL to reference issues. The issue you are referencing has no explanation of these examples. It would be better to explain them in https://github.com/sympy/sympy/issues/27975

Harikrishna-Srinivasan committed:
Corrected tests

Harikrishna-Srinivasan committed:
Added some more tests

oscarbenjamin commented on sympy/solvers/tests/test_recurr.py:
```diff
@@ -259,6 +274,13 @@ def test_issue_17990():
     e = sol.subs({C0: 1, C1: 1, C2: 1, n: 1}).evalf()
     assert abs(e + 0.130434782608696) < 1e-13
 
+@XFAIL
+def test_issue_27975():
+    assert rsolve_hyper([-2*2**n, 2*2**n], -2*binomial(n, k) + binomial(n + 1, k), k) == \
+        C0 - gamma(n + 1)/(2*2**n*gamma(k)*gamma(-k + n + 2))
+
+    term = -2**(1 - k)*factorial(k + 2)/factorial(k - 1) + factorial(k + 3)/(2**k*factorial(k))
+    assert rsolve_hyper([-2**n, 2**n], term, k) == 2**(-k - n + 1)*k*(k + 1)*(k + 2) + C0
```
An XFAIL test should test exactly one thing so that when that one thing starts working the test will XPASS.

Harikrishna-Srinivasan replied:
I checked both tests with your my code, & both passed. And, both are related to same issue only

oscarbenjamin replied:
If we are going to add these XFAIL tests then it needs to be split into two functions.

Also an XFAIL test needs to be written quite carefully so that it would still pass if any equivalent form was returned by the functions so perhaps:
```
assert rsolve_hyper(...) is not None
```

Harikrishna-Srinivasan replied:
What about `.equals()` will that be okay. Because we cannot accept any non-None values.

Harikrishna-Srinivasan committed:
XFAIL changed to not None

Harikrishna-Srinivasan committed:
Removed unused import

oscarbenjamin commented on sympy/solvers/tests/test_recurr.py:
```diff
@@ -259,6 +273,12 @@ def test_issue_17990():
     e = sol.subs({C0: 1, C1: 1, C2: 1, n: 1}).evalf()
     assert abs(e + 0.130434782608696) < 1e-13
 
+@XFAIL
+def test_issue_27975():
+    assert rsolve_hyper([-2*2**n, 2*2**n], -2*binomial(n, k) + binomial(n + 1, k), k) is not None
+
+    term = -2**(1 - k)*factorial(k + 2)/factorial(k - 1) + factorial(k + 3)/(2**k*factorial(k))
+    assert rsolve_hyper([-2**n, 2**n], term, k) is not None
```
It needs to be two separate functions.

Harikrishna-Srinivasan committed:
Separated the tests

Harikrishna-Srinivasan committed:
Merge branch 'sympy:master' into master

oscarbenjamin commented:
Looks good. Thanks.
